File: server/app/api/v1/morphology.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.core.dependencies import require_reception

from app.schemas.morphology import (
    MorphologyCreate,
    MorphologyUpdate,
    MorphologyResponse,
)

from app.services.morphology import MorphologyService

router = APIRouter(
    prefix="/morphologies",
    tags=["Morphologies"],
)


@router.post(
    "",
    response_model=MorphologyResponse,
)
def create_morphology(
    data: MorphologyCreate,
    db: Session = Depends(get_db),
    _: dict = Depends(require_reception),
):
    return MorphologyService.create_morphology(
        db,
        data,
    )


@router.get(
    "",
    response_model=list[MorphologyResponse],
)
def get_morphologies(
    db: Session = Depends(get_db),
    _: dict = Depends(require_reception),
):
    return MorphologyService.get_morphologies(db)


@router.get(
    "/{sample_code}",
    response_model=MorphologyResponse,
)
def get_morphology(
    sample_code: str,
    db: Session = Depends(get_db),
    _: dict = Depends(require_reception),
):
    return MorphologyService.get_morphology(
        db,
        sample_code,
    )


@router.patch(
    "/{sample_code}",
    response_model=MorphologyResponse,
)
def update_morphology(
    sample_code: str,
    data: MorphologyUpdate,
    db: Session = Depends(get_db),
    _: dict = Depends(require_reception),
):
    return MorphologyService.update_morphology(
        db,
        sample_code,
        data,
    )

File: client/src/api/morphology.ts
import api from "@/services/api";

import type {
    Morphology,
    MorphologyFormData,
} from "@/lib/schemas/morphology";

export async function getMorphology(
    sampleCode: string
): Promise<Morphology> {
    const response = await api.get(
        `/morphologies/${sampleCode}`
    );

    return response.data;
}

export async function createMorphology(
    data: MorphologyFormData
): Promise<Morphology> {
    const response = await api.post(
        "/morphologies",
        data
    );

    return response.data;
}

export async function updateMorphology(
    sampleCode: string,
    data: MorphologyFormData
): Promise<Morphology> {
    const response = await api.patch(
        `/morphologies/${sampleCode}`,
        data
    );

    return response.data;
}